x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/ExpressionBuilder.java

```diff
@@ -709,24 +756,31 @@ public List<NamedExpression> visitGrouping(EsqlBaseParser.FieldsContext ctx) {
     }
 
     @Override
-    public Object visitInputParam(EsqlBaseParser.InputParamContext ctx) {
+    public Expression visitInputParam(EsqlBaseParser.InputParamContext ctx) {
         QueryParam param = paramByToken(ctx.PARAM());
         return visitParam(ctx, param);
     }
 
     @Override
-    public Object visitInputNamedOrPositionalParam(EsqlBaseParser.InputNamedOrPositionalParamContext ctx) {
+    public Expression visitInputNamedOrPositionalParam(EsqlBaseParser.InputNamedOrPositionalParamContext ctx) {
         QueryParam param = paramByNameOrPosition(ctx.NAMED_OR_POSITIONAL_PARAM());
         if (param == null) {
             return Literal.NULL;
         }
         return visitParam(ctx, param);
     }
 
-    private Object visitParam(EsqlBaseParser.ParamsContext ctx, QueryParam param) {
+    private Expression visitParam(EsqlBaseParser.ParamsContext ctx, QueryParam param) {
         Source source = source(ctx);
         DataType type = param.type();
-        return new Literal(source, param.value(), type);
+        if (param.isPattern()) {
+            // let visitQualifiedNamePattern create a real UnresolvedNamePattern with Automaton
+            return new UnresolvedNamePattern(source, null, (String) param.value(), (String) param.value());
```

Comment:
Instead of casting, store param.value() into a variable which then gets consumed inside the if - instead of casting do a .toString().
Also consider the null case.